Distribution/Server/Packages/Unpack.hs

```diff
@@ -178,25 +200,72 @@ basicChecks lax now tarGzFile contents = do
 portableTakeFileName :: FilePath -> String
 portableTakeFileName = System.FilePath.Windows.takeFileName
 
+tarOps :: PackageIdentifier -> TarIndex -> CheckPackageContentOps UploadMonad
+tarOps pkgId tarIndex = CheckPackageContentOps {
+  doesFileExist        = fileExist    . relative,
+  doesDirectoryExist   = dirExist . relative . System.FilePath.Posix.addTrailingPathSeparator,
+  getDirectoryContents = dirContents . System.FilePath.Posix.dropTrailingPathSeparator . relative,
+  getFileContents      = fileContents . relative
+}
+  where
+    -- The tar index has names like <pkgid>/foo.cabal, but the
+    -- CheckPackageContentOps requests files without specifying the pkgid
+    -- root. We convert the requested file paths into the tar index format.
+    relative = normalise . (display pkgId </>)
+    -- Build the map. In case of multiple intries for a file, we want the
+    -- first entry in the tar file to win. Since the tarIndex is the reversed
+    -- tar file, we need to reverse it back.
+    fileMap = Map.fromList (reverse tarIndex)
```

Comment:
The comment here is wrong, but actually I think the code is right.

In case of multiple entries for a file we actually want the **last** entry to win (since this is tar file append-style semantics). Since the tarIndex is the reversed tar file and Map.fromList makes the last one win, we need to reverse it back.